internal/webhook/external/quay.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	QuaySecretDataKey = "secret"
+
+	quay                    = "quay"
+	quayWebhookBodyMaxBytes = 2 << 20 // 2MB
+)
+
+func init() {
+	registry.register(
+		quay,
+		webhookReceiverRegistration{
+			predicate: func(cfg kargoapi.WebhookReceiverConfig) bool {
+				return cfg.Quay != nil
+			},
+			factory: newQuayWebhookReceiver,
+		},
+	)
+}
+
+// quayWebhookReceiver is an implementation of WebhookReceiver that handles
+// inbound webhooks from Quay.
+type quayWebhookReceiver struct {
+	*baseWebhookReceiver
+}
+
+// newQuayWebhookReceiver returns a new instance of quayWebhookReceiver.
+func newQuayWebhookReceiver(
+	c client.Client,
+	project string,
+	cfg kargoapi.WebhookReceiverConfig,
+) WebhookReceiver {
+	return &quayWebhookReceiver{
+		baseWebhookReceiver: &baseWebhookReceiver{
+			client:     c,
+			project:    project,
+			secretName: cfg.Quay.SecretRef.Name,
+		},
+	}
+}
+
+// GetDetails implements WebhookReceiver.
+func (q *quayWebhookReceiver) getReceiverType() string {
+	return quay
+}
+
+// getSecretValues implements WebhookReceiver.
+func (q *quayWebhookReceiver) getSecretValues(
+	secretData map[string][]byte,
+) ([]string, error) {
+	secretValue, ok := secretData[QuaySecretDataKey]
+	if !ok {
+		return nil,
+			errors.New("Secret data is not valid for a Quay WebhookReceiver")
+	}
+	return []string{string(secretValue)}, nil
+}
+
+func (q *quayWebhookReceiver) GetHandler() http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx).WithValues("path", r.URL.Path)
+		ctx = logging.ContextWithLogger(ctx, logger)
+		logger.Debug("identifying source repository")
+
+		const maxBytes = 2 << 20 // 2MB
+		b, err := io.LimitRead(r.Body, maxBytes)
+		if err != nil {
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					fmt.Errorf("failed to read request body: %w", err),
+					http.StatusRequestEntityTooLarge,
+				),
+			)
+			return
+		}
+
+		payload := struct {
+			// format: quay.io/mynamespace/repository
+			RepoWebURL string `json:"docker_url"`
+		}{}
+
+		if err = json.Unmarshal(b, &payload); err != nil {
+			logger.Error(err, "failed to unmarshal request body")
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					fmt.Errorf("failed to unmarshal request body: %w", err),
+					http.StatusBadRequest,
+				),
+			)
+			return
+		}
+
+		if payload.RepoWebURL == "" {
+			logger.Debug("missing repository web URL in request body")
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(
+					fmt.Errorf("missing repository web URL in request body"),
+					http.StatusBadRequest,
+				),
+			)
+			return
+		}
+
+		logger = logger.WithValues("repoWebURL", payload.RepoWebURL)
+		ctx = logging.ContextWithLogger(ctx, logger)
+		result, err := refreshWarehouses(ctx, q.client, q.project, payload.RepoWebURL)
+		if err != nil {
+			xhttp.WriteErrorJSON(w,
+				xhttp.Error(err, http.StatusInternalServerError),
+			)
```

Comment:
The behavior of `xhttp.WriteErrorJSON` is already to treat everything that isn't otherwise qualified as a `http.StatusInternalServerError`, so all cases of an unclassified error do not require any special wrapping.

```suggestion
			xhttp.WriteErrorJSON(w, err)
```